File: src/server/protocol/lspProt.ml
(*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

type client_id = int

type error_kind =
  | ExpectedError
  | UnexpectedError

type error_info = error_kind * string * Utils.callstack

type metadata = {
  (* when did this work-item get triggered? *)
  start_wall_time: float;
  (* What was the thing that triggered this work-item *)
  start_json_truncated: Hh_json.json;
  (* What was the state of the server at the time the work-item was triggered? *)
  (* Might be None e.g. if the server was down at the time or if we don't know *)
  start_server_status: ServerStatus.status option;
  start_watcher_status: FileWatcherStatus.status option;
  (* And what was the state of the lspCommand client? Is optional only to save *)
  (* space in the obvious cases that don't need explanation. *)
  start_lsp_state: string option;
  start_lsp_state_reason: string option;
  (* If handling the workitem resulted in error, what was that error? *)
  error_info: error_info option;
  (* If the workitem was handled on the server, how long did it take there? *)
  server_profiling: Profiling_js.finished option;
  (* and if it had work done on the client, how long there? *)
  client_duration: float option;
  (* Did the handler for this workitem provide any extra data? *)
  extra_data: (string * Hh_json.json) list;
  (* The logging context for the server *)
  server_logging_context: FlowEventLogger.logging_context option;
  (* LSP method (e.g. 'textDocument/completion') *)
  lsp_method_name: string;
  (* If we're tracking an interaction in the lsp process, this is the id of the interaction *)
  interaction_tracking_id: int option;
}

(** For LSP work-items, we keep metadata about requests, to help us log better telemetry.
    After the work has been handled, we fill out the second part of the metadata.
 *)
let empty_metadata =
  {
    start_wall_time = 0.0;
    start_server_status = None;
    start_watcher_status = None;
    start_json_truncated = Hh_json.JSON_Object [];
    start_lsp_state = None;
    start_lsp_state_reason = None;
    error_info = None;
    server_profiling = None;
    client_duration = None;
    extra_data = [];
    server_logging_context = None;
    lsp_method_name = "";
    interaction_tracking_id = None;
  }

(* This is the reason why we start to do a recheck. Since rechecks can be combined together, there
 * may be multiple reasons for a single recheck *)
type recheck_reason =
  (* One file changed on disk. *)
  | Single_file_changed of { filename: string }
  (* More than one file changed on disk *)
  | Many_files_changed of { file_count: int }
  (* If we're using Watchman as the filewatcher, we can tell when the mergebase changed.
   * We can differentiate that from Many_files_changed *)
  | Rebased of { file_count: int }
  (* If watchman restarts, it may have missed changes. We can recheck everything to
     get back on track. *)
  | File_watcher_missed_changes
  (* If try to autocomplete in foo.js and it's dependencies are unchecked, then we start a recheck
   * with a reason of Unchecked_dependencies { filename = "/path/to/foo.js"; } *)
  | Unchecked_dependencies of { filename: string }
  (* A lazy server started from saved state has an old dependency graph and has to update it *)
  | Lazy_init_update_deps
  (* A lazy server may decided to typecheck some files during init (like Watchman lazy mode will
   * typecheck files which have changed since the mergebase) *)
  | Lazy_init_typecheck
  (* At init when we do a full check *)
  | Full_init

let verbose_string_of_recheck_reason = function
  | Single_file_changed { filename } -> Printf.sprintf "1 file changed (%s)" filename
  | Many_files_changed { file_count } -> Printf.sprintf "%d files changed" file_count
  | Rebased { file_count } -> Printf.sprintf "Rebased (%d files changed)" file_count
  | File_watcher_missed_changes -> "Resynchronizing file watcher"
  | Unchecked_dependencies { filename } -> Printf.sprintf "Unchecked dependencies of %s" filename
  | Lazy_init_update_deps -> "Lazy init update deps"
  | Lazy_init_typecheck -> "Lazy init typecheck"
  | Full_init -> "Full init"

let normalized_string_of_recheck_reason = function
  | Single_file_changed { filename = _ } -> "singleFileChanged"
  | Many_files_changed { file_count = _ } -> "manyFilesChanged"
  | Rebased { file_count = _ } -> "rebased"
  | File_watcher_missed_changes -> "fileWatcherMissedChanges"
  | Unchecked_dependencies { filename = _ } -> "uncheckedDependencies"
  | Lazy_init_update_deps -> "lazyInitUpdateDeps"
  | Lazy_init_typecheck -> "lazyInitTypecheck"
  | Full_init -> "fullInit"

type request =
  | Subscribe
  | LspToServer of Lsp.lsp_message
  | LiveErrorsRequest of Lsp.DocumentUri.t

type request_with_metadata = request * metadata

(* requests, notifications, responses from client *)

let string_of_request = function
  | Subscribe -> "subscribe"
  | LspToServer msg -> Printf.sprintf "lspToServer %s" (Lsp_fmt.message_name_to_string msg)
  | LiveErrorsRequest uri -> Printf.sprintf "liveErrorsRequest %s" (Lsp.DocumentUri.to_string uri)

let string_of_request_with_metadata (request, _) = string_of_request request

let json_of_request =
  Hh_json.(
    function
    | (Subscribe, _) -> JSON_Object [("method", JSON_String "subscribe")]
    | (LspToServer _, metadata) -> metadata.start_json_truncated
    | (LiveErrorsRequest uri, metadata) ->
      JSON_Object
        [
          ("method", JSON_String "liveErrorsRequest");
          ("params", JSON_Object [("uri", JSON_String (Lsp.DocumentUri.to_string uri))]);
          ("trigger", metadata.start_json_truncated);
        ])

(* Why is the server sending us a list of errors *)
type errors_reason =
  (* Sending all the errors at the end of the recheck *)
  | End_of_recheck of { recheck_reasons: recheck_reason list }
  (* Streaming errors during recheck *)
  | Recheck_streaming of { recheck_reasons: recheck_reason list }
  (* Sometimes the env changes, which influences which errors we send to the lsp. For example, we
   * only send warnings for open files. When a file is opened or closed, we have to recalculate
   * which warnings to send and send the updated set. *)
  | Env_change
  (* The persistent client just subscribed to errors, so was sent the initial error list *)
  | New_subscription

type error_response_kind =
  | Canceled_error_response
  | Errored_error_response

type live_errors_failure = {
  live_errors_failure_kind: error_response_kind;
  live_errors_failure_reason: string;
  live_errors_failure_uri: Lsp.DocumentUri.t;
}

type live_errors_response = {
  live_diagnostics: Lsp.PublishDiagnostics.diagnostic list;
  live_errors_uri: Lsp.DocumentUri.t;
}

type response =
  | LspFromServer of Lsp.lsp_message option
  | LiveErrorsResponse of (live_errors_response, live_errors_failure) result
  | UncaughtException of {
      request: request;
      exception_constructor: string;
      stack: string;
    }

type response_with_metadata = response * metadata

type notification_from_server =
  | Errors of {
      diagnostics: Lsp.PublishDiagnostics.diagnostic list Lsp.UriMap.t;
      errors_reason: errors_reason;
    }
  | StartRecheck
  | EndRecheck of ServerProt.Response.lazy_stats
  | ServerExit of Exit.t  (** only used for the subset of exits which client handles *)
  | Please_hold of (ServerStatus.status * FileWatcherStatus.status)
  | EOF  (** monitor is about to close the connection *)

type message_from_server =
  | RequestResponse of response_with_metadata
  | NotificationFromServer of notification_from_server

let string_of_response = function
  | LspFromServer None -> "lspFromServer None"
  | LspFromServer (Some msg) ->
    Printf.sprintf "lspFromServer %s" (Lsp_fmt.message_name_to_string msg)
  | LiveErrorsResponse (Ok { live_diagnostics; live_errors_uri; _ }) ->
    let (errors, warnings, others) =
      Base.List.fold_left
        ~f:(fun (errors, warnings, others) { Lsp.PublishDiagnostics.severity; _ } ->
          match severity with
          | Some Lsp.PublishDiagnostics.Error -> (errors + 1, warnings, others)
          | Some Lsp.PublishDiagnostics.Warning -> (errors, warnings + 1, others)
          | Some Lsp.PublishDiagnostics.Information
          | Some Lsp.PublishDiagnostics.Hint
          | None ->
            (errors, warnings, others + 1))
        ~init:(0, 0, 0)
        live_diagnostics
    in
    Printf.sprintf
      "liveErrorsResponse OK (%d errors, %d warnings, %d other) %s"
      errors
      warnings
      others
      (Lsp.DocumentUri.to_string live_errors_uri)
  | LiveErrorsResponse
      (Error { live_errors_failure_kind; live_errors_failure_reason; live_errors_failure_uri }) ->
    Printf.sprintf
      "liveErrorsResponse %s %s %s"
      (match live_errors_failure_kind with
      | Canceled_error_response -> "CANCELED"
      | Errored_error_response -> "ERRORED")
      live_errors_failure_reason
      (Lsp.DocumentUri.to_string live_errors_failure_uri)
  | UncaughtException { request; exception_constructor; stack } ->
    Printf.sprintf
      "UncaughtException %s in handling `%s`: %s"
      exception_constructor
      (string_of_request request)
      stack

let string_of_message_from_server = function
  | RequestResponse (response, _) -> string_of_response response
  | NotificationFromServer notification ->
    begin
      match notification with
      | Errors _ -> "errors"
      | StartRecheck -> "startRecheck"
      | EndRecheck _ -> "endRecheck"
      | ServerExit code -> "serverExit_" ^ Exit.to_string code
      | Please_hold (server_status, watcher_status) ->
        Printf.sprintf
          "pleaseHold_server=%s_watcher=%s"
          (ServerStatus.string_of_status server_status)
          (FileWatcherStatus.string_of_status watcher_status)
      | EOF -> "EOF"
    end

type message_from_server_mapper = {
  of_live_errors_failure: message_from_server_mapper -> live_errors_failure -> live_errors_failure;
  of_live_errors_response:
    message_from_server_mapper -> live_errors_response -> live_errors_response;
  of_message_from_server: message_from_server_mapper -> message_from_server -> message_from_server;
  of_notification:
    message_from_server_mapper -> notification_from_server -> notification_from_server;
  of_response: message_from_server_mapper -> response -> response;
}

let default_message_from_server_mapper ~(lsp_mapper : Lsp_mapper.t) =
  let open Lsp_mapper in
  {
    of_live_errors_failure =
      (fun _mapper { live_errors_failure_kind; live_errors_failure_reason; live_errors_failure_uri } ->
        let live_errors_failure_uri =
          lsp_mapper.of_document_uri lsp_mapper live_errors_failure_uri
        in
        { live_errors_failure_kind; live_errors_failure_reason; live_errors_failure_uri });
    of_live_errors_response =
      (fun _mapper { live_diagnostics; live_errors_uri } ->
        let live_diagnostics =
          Base.List.map ~f:(lsp_mapper.of_diagnostic lsp_mapper) live_diagnostics
        in
        let live_errors_uri = lsp_mapper.of_document_uri lsp_mapper live_errors_uri in
        { live_diagnostics; live_errors_uri });
    of_message_from_server =
      (fun mapper message ->
        match message with
        | RequestResponse (response, metadata) ->
          RequestResponse (mapper.of_response mapper response, metadata)
        | NotificationFromServer notification ->
          NotificationFromServer (mapper.of_notification mapper notification));
    of_notification =
      (fun _mapper notif ->
        match notif with
        | Errors { diagnostics; errors_reason } ->
          let diagnostics =
            Lsp.UriMap.fold
              (fun uri diags acc ->
                let uri = lsp_mapper.of_document_uri lsp_mapper uri in
                let diags = Base.List.map ~f:(lsp_mapper.of_diagnostic lsp_mapper) diags in
                Lsp.UriMap.add uri diags acc)
              diagnostics
              Lsp.UriMap.empty
          in
          Errors { diagnostics; errors_reason }
        | StartRecheck -> StartRecheck
        | EndRecheck stats -> EndRecheck stats
        | ServerExit exit_status -> ServerExit exit_status
        | Please_hold (server_status, file_watcher_status) ->
          Please_hold (server_status, file_watcher_status)
        | EOF -> EOF);
    of_response =
      (fun mapper response ->
        match response with
        | LspFromServer lsp ->
          LspFromServer (Base.Option.map ~f:(lsp_mapper.of_lsp_message lsp_mapper) lsp)
        | LiveErrorsResponse (Ok live_errors_response) ->
          LiveErrorsResponse (Ok (mapper.of_live_errors_response mapper live_errors_response))
        | LiveErrorsResponse (Error live_errors_failure) ->
          LiveErrorsResponse (Error (mapper.of_live_errors_failure mapper live_errors_failure))
        | UncaughtException { request; exception_constructor; stack } ->
          UncaughtException { request; exception_constructor; stack });
  }

File: rust_port/crates/flow_monitor_rpc/src/lsp_prot.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use dupe::Dupe;
use lsp_types::Diagnostic;
use lsp_types::DiagnosticSeverity;

use crate::file_watcher_status;
use crate::server_status;

mod flow_exit_status_serde {
    use flow_common_exit_status::FlowExitStatus;
    use serde::Deserialize;
    use serde::Deserializer;
    use serde::Serialize;
    use serde::Serializer;

    pub fn serialize<S>(status: &FlowExitStatus, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        flow_common_exit_status::error_code(*status).serialize(serializer)
    }

    pub fn deserialize<'de, D>(deserializer: D) -> Result<FlowExitStatus, D::Error>
    where
        D: Deserializer<'de>,
    {
        let code = i32::deserialize(deserializer)?;
        match code {
            -6 => Ok(FlowExitStatus::Interrupted),
            0 => Ok(FlowExitStatus::NoError),
            1 => Ok(FlowExitStatus::WindowsKilledByTaskManager),
            2 => Ok(FlowExitStatus::TypeError),
            3 => Ok(FlowExitStatus::OutOfTime),
            4 => Ok(FlowExitStatus::KillError),
            5 => Ok(FlowExitStatus::UnusedServer),
            6 => Ok(FlowExitStatus::NoServerRunning),
            7 => Ok(FlowExitStatus::OutOfRetries),
            8 => Ok(FlowExitStatus::InvalidFlowconfig),
            9 => Ok(FlowExitStatus::BuildIdMismatch),
            10 => Ok(FlowExitStatus::InputError),
            11 => Ok(FlowExitStatus::LockStolen),
            12 => Ok(FlowExitStatus::CouldNotFindFlowconfig),
            13 => Ok(FlowExitStatus::ServerOutOfDate),
            15 => Ok(FlowExitStatus::OutOfSharedMemory),
            16 => Ok(FlowExitStatus::FlowconfigChanged),
            17 => Ok(FlowExitStatus::PathIsNotAFile),
            18 => Ok(FlowExitStatus::Autostop),
            19 => Ok(FlowExitStatus::KilledByMonitor),
            20 => Ok(FlowExitStatus::InvalidSavedState),
            21 => Ok(FlowExitStatus::Restart),
            22 => Ok(FlowExitStatus::CouldNotExtractFlowlibs),
            64 => Ok(FlowExitStatus::CommandlineUsageError),
            66 => Ok(FlowExitStatus::NoInput),
            78 => Ok(FlowExitStatus::ServerStartFailed),
            97 => Ok(FlowExitStatus::MissingFlowlib),
            98 => Ok(FlowExitStatus::SocketError),
            99 => Ok(FlowExitStatus::DfindDied),
            101 => Ok(FlowExitStatus::WatchmanError),
            102 => Ok(FlowExitStatus::HashTableFull),
            103 => Ok(FlowExitStatus::HeapFull),
            104 => Ok(FlowExitStatus::WatchmanFailed),
            105 => Ok(FlowExitStatus::FileWatcherMissedChanges),
            108 => Ok(FlowExitStatus::EventLoggerRestartOutOfRetries),
            110 => Ok(FlowExitStatus::UnknownError),
            111 => Ok(FlowExitStatus::EdenfsWatcherFailed),
            112 => Ok(FlowExitStatus::EdenfsWatcherLostChanges),
            _ => Err(serde::de::Error::custom(format!(
                "unknown FlowExitStatus code {}",
                code
            ))),
        }
    }
}

pub type ClientId = i32;

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub enum ErrorKind {
    ExpectedError,
    UnexpectedError,
}

pub type ErrorInfo = (ErrorKind, String, String);

pub type Json = serde_json::Value;

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct LoggingContext {
    pub from: Option<String>,
    pub agent_id: Option<String>,
}

pub type ProfilingFinished = serde_json::Value;

pub type LspId = flow_lsp::lsp::LspId;

pub type LspMessage = flow_lsp::lsp::LspMessage;

pub type DocumentUri = flow_lsp::lsp::DocumentUri;

pub type UriMap<V> = flow_lsp::lsp::UriMap<V>;

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct Metadata {
    // when did this work-item get triggered?
    pub start_wall_time: f64,
    // What was the thing that triggered this work-item
    pub start_json_truncated: Json,
    // What was the state of the server at the time the work-item was triggered?
    // Might be None e.g. if the server was down at the time or if we don't know
    pub start_server_status: Option<server_status::Status>,
    pub start_watcher_status: Option<file_watcher_status::Status>,
    // And what was the state of the lspCommand client? Is optional only to save
    // space in the obvious cases that don't need explanation.
    pub start_lsp_state: Option<String>,
    pub start_lsp_state_reason: Option<String>,
    // If handling the workitem resulted in error, what was that error?
    pub error_info: Option<ErrorInfo>,
    // If the workitem was handled on the server, how long did it take there?
    pub server_profiling: Option<ProfilingFinished>,
    // and if it had work done on the client, how long there?
    pub client_duration: Option<f64>,
    // Did the handler for this workitem provide any extra data?
    pub extra_data: Vec<(String, Json)>,
    // The logging context for the server
    pub server_logging_context: Option<LoggingContext>,
    // LSP method (e.g. 'textDocument/completion')
    pub lsp_method_name: String,
    // If we're tracking an interaction in the lsp process, this is the id of the interaction
    pub interaction_tracking_id: Option<i32>,
    // JSON-RPC id of corresponding LSP request
    pub lsp_id: Option<LspId>,
    pub activity_key: Option<Json>,
}

pub fn empty_metadata() -> Metadata {
    Metadata {
        start_wall_time: 0.0,
        start_server_status: None,
        start_watcher_status: None,
        start_json_truncated: serde_json::Value::Object(Default::default()),
        start_lsp_state: None,
        start_lsp_state_reason: None,
        error_info: None,
        server_profiling: None,
        client_duration: None,
        extra_data: Vec::new(),
        server_logging_context: None,
        lsp_method_name: String::new(),
        interaction_tracking_id: None,
        lsp_id: None,
        activity_key: None,
    }
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub enum Request {
    Subscribe,
    LspToServer(LspMessage),
    LiveErrorsRequest(DocumentUri),
}

pub type RequestWithMetadata = (Request, Metadata);

// requests, notifications, responses from client

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub enum ErrorsReason {
    // Sending all the errors at the end of the recheck
    EndOfRecheck,
    // Streaming errors during recheck
    RecheckStreaming,
    EnvChange,
    NewSubscription,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub enum ErrorResponseKind {
    CanceledErrorResponse,
    ErroredErrorResponse,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct LiveErrorsFailure {
    pub live_errors_failure_kind: ErrorResponseKind,
    pub live_errors_failure_reason: String,
    pub live_errors_failure_uri: DocumentUri,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct LiveErrorsResponse {
    pub live_diagnostics: Vec<Diagnostic>,
    pub live_errors_uri: DocumentUri,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub enum Response {
    LspFromServer(Option<LspMessage>),
    LiveErrorsResponse(Result<LiveErrorsResponse, LiveErrorsFailure>),
    UncaughtException {
        request: Request,
        exception_constructor: String,
        stack: String,
    },
}

pub type ResponseWithMetadata = (Response, Metadata);

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct RecheckStats {
    pub dependent_file_count: i32,
    pub changed_file_count: i32,
    // name of cycle leader, and size of cycle
    pub top_cycle: Option<(flow_parser::file_key::FileKey, i32)>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub enum TelemetryFromServer {
    InitSummary { duration: f64 },
    RecheckSummary { stats: RecheckStats, duration: f64 },
    CommandSummary { name: String, duration: f64 },
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub enum NotificationFromServer {
    Errors {
        diagnostics: UriMap<Vec<Diagnostic>>,
        errors_reason: ErrorsReason,
    },
    StartRecheck,
    EndRecheck(crate::server_prot::response::LazyStats),
    // only used for the subset of exits which client handles
    ServerExit(#[serde(with = "flow_exit_status_serde")] flow_common_exit_status::FlowExitStatus),
    PleaseHold(server_status::Status, file_watcher_status::Status),
    Telemetry(TelemetryFromServer),
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub enum MessageFromServer {
    RequestResponse(ResponseWithMetadata),
    NotificationFromServer(NotificationFromServer),
}

pub fn string_of_request(request: &Request) -> String {
    match request {
        Request::Subscribe => "subscribe".to_string(),
        Request::LspToServer(msg) => format!("lspToServer {:?}", msg),
        Request::LiveErrorsRequest(uri) => format!("liveErrorsRequest {}", uri),
    }
}

pub fn string_of_request_with_metadata((request, _): &RequestWithMetadata) -> String {
    string_of_request(request)
}

pub fn json_of_request((request, metadata): &RequestWithMetadata) -> serde_json::Value {
    match request {
        Request::Subscribe => {
            serde_json::json!({ "method": "subscribe" })
        }
        Request::LspToServer(_) => metadata.start_json_truncated.clone(),
        Request::LiveErrorsRequest(uri) => {
            serde_json::json!({
                "method": "liveErrorsRequest",
                "params": { "uri": uri.to_string() },
                "trigger": metadata.start_json_truncated,
            })
        }
    }
}

pub fn string_of_response(response: &Response) -> String {
    match response {
        Response::LspFromServer(None) => "lspFromServer None".to_string(),
        Response::LspFromServer(Some(msg)) => format!("lspFromServer {:?}", msg),
        Response::LiveErrorsResponse(Ok(LiveErrorsResponse {
            live_diagnostics,
            live_errors_uri,
        })) => {
            let (errors, warnings, others) =
                live_diagnostics
                    .iter()
                    .fold((0, 0, 0), |(errors, warnings, others), diag| {
                        match diag.severity {
                            Some(s) if s == DiagnosticSeverity::ERROR => {
                                (errors + 1, warnings, others)
                            }
                            Some(s) if s == DiagnosticSeverity::WARNING => {
                                (errors, warnings + 1, others)
                            }
                            Some(_) | None => (errors, warnings, others + 1),
                        }
                    });
            format!(
                "liveErrorsResponse OK ({} errors, {} warnings, {} other) {}",
                errors, warnings, others, live_errors_uri
            )
        }
        Response::LiveErrorsResponse(Err(LiveErrorsFailure {
            live_errors_failure_kind,
            live_errors_failure_reason,
            live_errors_failure_uri,
        })) => {
            format!(
                "liveErrorsResponse {} {} {}",
                match live_errors_failure_kind {
                    ErrorResponseKind::CanceledErrorResponse => "CANCELED",
                    ErrorResponseKind::ErroredErrorResponse => "ERRORED",
                },
                live_errors_failure_reason,
                live_errors_failure_uri
            )
        }
        Response::UncaughtException {
            request,
            exception_constructor,
            stack,
        } => {
            format!(
                "UncaughtException {} in handling `{}`: {}",
                exception_constructor,
                string_of_request(request),
                stack
            )
        }
    }
}

pub struct MessageFromServerMapper {
    pub of_live_errors_failure:
        Box<dyn Fn(&MessageFromServerMapper, LiveErrorsFailure) -> LiveErrorsFailure>,
    pub of_live_errors_response:
        Box<dyn Fn(&MessageFromServerMapper, LiveErrorsResponse) -> LiveErrorsResponse>,
    pub of_message_from_server:
        Box<dyn Fn(&MessageFromServerMapper, MessageFromServer) -> MessageFromServer>,
    pub of_notification:
        Box<dyn Fn(&MessageFromServerMapper, NotificationFromServer) -> NotificationFromServer>,
    pub of_response: Box<dyn Fn(&MessageFromServerMapper, Response) -> Response>,
}

pub fn default_message_from_server_mapper(
    lsp_mapper: std::sync::Arc<flow_lsp::lsp_mapper::T>,
) -> MessageFromServerMapper {
    let lsp_mapper_clone1 = lsp_mapper.clone();
    let of_live_errors_failure = Box::new(
        move |_mapper: &MessageFromServerMapper,
              LiveErrorsFailure {
                  live_errors_failure_kind,
                  live_errors_failure_reason,
                  live_errors_failure_uri,
              }: LiveErrorsFailure| {
            let live_errors_failure_uri =
                (lsp_mapper_clone1.of_document_uri)(&lsp_mapper_clone1, live_errors_failure_uri);
            LiveErrorsFailure {
                live_errors_failure_kind,
                live_errors_failure_reason,
                live_errors_failure_uri,
            }
        },
    );

    let lsp_mapper_clone2 = lsp_mapper.dupe();
    let of_live_errors_response = Box::new(
        move |_mapper: &MessageFromServerMapper,
              LiveErrorsResponse {
                  live_diagnostics,
                  live_errors_uri,
              }: LiveErrorsResponse| {
            let live_diagnostics = live_diagnostics
                .into_iter()
                .map(|d| (lsp_mapper_clone2.of_diagnostic)(&lsp_mapper_clone2, d))
                .collect();
            let live_errors_uri =
                (lsp_mapper_clone2.of_document_uri)(&lsp_mapper_clone2, live_errors_uri);
            LiveErrorsResponse {
                live_diagnostics,
                live_errors_uri,
            }
        },
    );

    let of_message_from_server = Box::new(
        move |mapper: &MessageFromServerMapper, message: MessageFromServer| match message {
            MessageFromServer::RequestResponse((response, metadata)) => {
                MessageFromServer::RequestResponse((
                    (mapper.of_response)(mapper, response),
                    metadata,
                ))
            }
            MessageFromServer::NotificationFromServer(notification) => {
                MessageFromServer::NotificationFromServer((mapper.of_notification)(
                    mapper,
                    notification,
                ))
            }
        },
    );

    let lsp_mapper_clone3 = lsp_mapper.clone();
    let of_notification = Box::new(
        move |_mapper: &MessageFromServerMapper, notif: NotificationFromServer| match notif {
            NotificationFromServer::Errors {
                diagnostics,
                errors_reason,
            } => {
                let diagnostics =
                    diagnostics
                        .into_iter()
                        .fold(UriMap::new(), |mut acc, (uri, diags)| {
                            let uri = (lsp_mapper_clone3.of_document_uri)(&lsp_mapper_clone3, uri);
                            let diags = diags
                                .into_iter()
                                .map(|d| (lsp_mapper_clone3.of_diagnostic)(&lsp_mapper_clone3, d))
                                .collect();
                            acc.insert(uri, diags);
                            acc
                        });
                NotificationFromServer::Errors {
                    diagnostics,
                    errors_reason,
                }
            }
            NotificationFromServer::StartRecheck => NotificationFromServer::StartRecheck,
            NotificationFromServer::EndRecheck(stats) => NotificationFromServer::EndRecheck(stats),
            NotificationFromServer::ServerExit(exit_status) => {
                NotificationFromServer::ServerExit(exit_status)
            }
            NotificationFromServer::PleaseHold(server_status, file_watcher_status) => {
                NotificationFromServer::PleaseHold(server_status, file_watcher_status)
            }
            NotificationFromServer::Telemetry(t) => NotificationFromServer::Telemetry(t),
        },
    );

    let lsp_mapper_clone4 = lsp_mapper;
    let of_response = Box::new(
        move |mapper: &MessageFromServerMapper, response: Response| match response {
            Response::LspFromServer(lsp) => Response::LspFromServer(
                lsp.map(|msg| (lsp_mapper_clone4.of_lsp_message)(&lsp_mapper_clone4, msg)),
            ),
            Response::LiveErrorsResponse(Ok(live_errors_response)) => {
                Response::LiveErrorsResponse(Ok((mapper.of_live_errors_response)(
                    mapper,
                    live_errors_response,
                )))
            }
            Response::LiveErrorsResponse(Err(live_errors_failure)) => Response::LiveErrorsResponse(
                Err((mapper.of_live_errors_failure)(mapper, live_errors_failure)),
            ),
            Response::UncaughtException {
                request,
                exception_constructor,
                stack,
            } => Response::UncaughtException {
                request,
                exception_constructor,
                stack,
            },
        },
    );

    MessageFromServerMapper {
        of_live_errors_failure,
        of_live_errors_response,
        of_message_from_server,
        of_notification,
        of_response,
    }
}
